tasks/Task.cpp

```diff
@@ -2,48 +2,525 @@
+
+            if (holds_alternative<string>(data))
+            {
+                string data_string = get<string>(data);
+                vector<uint8_t> new_data(data_string.begin(), data_string.end());
+                dataPacket.data.resize(new_data.size());
+                dataPacket.data = new_data;
+            }
+            else
+            {
+                vector<byte> data_byte = get<binary>(data);
+                dataPacket.data.resize(data_byte.size());
+                for (unsigned int i = 0; i < data_byte.size(); i++)
+                {
+                    dataPacket.data[i] = to_integer<uint8_t>(data_byte[i]);
+                }
+            }
+            _data_out.write(dataPacket);
+        });
+}
+
+void Task::evaluateDataChannel()
+{
+    switch (mState.data_channel)
+    {
+    case DataChannelOpened:
+    {
+        iodrivers_base::RawPacket raw_packet;
+        if (_data_in.read(raw_packet) != RTT::NewData)
+            return;
+
+        vector<byte> data;
+        data.resize(raw_packet.data.size());
+        for (unsigned int i = 0; i < raw_packet.data.size(); i++)
+        {
+            data[i] = static_cast<byte>(raw_packet.data[i]);
+        }
+        mDataChannel->send(&data.front(), data.size());
+        break;
+    }
+    case DataChannelClosed:
+        throw runtime_error("DataChannel closed");
+    case DataChannelFailed:
+        throw runtime_error("DataChannel failed");
+    default:
+        break;
+    }
+}
+
+void Task::evaluateWebSocket()
+{
+    switch (mState.web_socket)
+    {
+    case WebSocketClosed:
+        throw runtime_error("WebSocket closed");
+    case WebSocketFailed:
+        throw runtime_error("WebSocket failed");
+    default:
+        break;
+    }
+}
+
+Task::Task(string const &name) : TaskBase(name) {}
+
+Task::~Task() {}
 
 /// The following lines are template definitions for the various state machine
 // hooks defined by Orocos::RTT. See Task.hpp for more detailed
 // documentation about them.
 
 bool Task::configureHook()
 {
-    if (! TaskBase::configureHook())
+    if (!TaskBase::configureHook())
         return false;
+
+    mConfig.iceServers.emplace_back(_stun_server.get());
+    mWs = make_shared<rtc::WebSocket>();
+
+    configureWebSocket();
+
     return true;
 }
 bool Task::startHook()
 {
-    if (! TaskBase::startHook())
+    if (!TaskBase::startHook())
         return false;
+
+    if (!_remote_peer_id.get().empty())
+    {
+        mRemotePeerID = _remote_peer_id.get();
+        mRemotePeerAnswerReceived = false;
+        future<void> wait_remote_peer_future = mWaitRemotePeerPromise.get_future();
+        // Try to get contact with the remote peer and create datachannel
+        ping();
+        future_status status =
+            wait_remote_peer_future.wait_until(chrono::system_clock::now() +
+                                               chrono::seconds(static_cast<int>(_wait_remote_peer_time_out.get().toSeconds())));
```

Comment:
Here, the while loop does make sense. It ensures that if we are configuring/starting two components at roughly the same time, they have a chance to find each other.

Reply:
Also, same comments about seconds and wait_for